crates/flox/src/commands/auth.rs

```diff
@@ -0,0 +1,295 @@
+                    "slow_down" => {
+                        if let DeviceFlowState::Processing(current_interval) = self.state {
+                            self.state =
+                                DeviceFlowState::Processing(current_interval + FIVE_SECONDS);
+                        };
+                    },
+                    other_reason => {
+                        self.state = DeviceFlowState::Failure(credential_error(format!(
+                            "Error checking for token: {}",
+                            other_reason
+                        )));
+                    },
+                }
+            } else {
+                let mut this_credential = Credential::empty();
+                this_credential.token = res["access_token"].as_str().unwrap().to_string();
+
+                if let Some(expires_in) = res.get("expires_in") {
+                    this_credential.expiry = calculate_expiry(expires_in.as_i64().unwrap());
+                    this_credential.refresh_token =
+                        res["refresh_token"].as_str().unwrap().to_string();
+                }
+
+                self.state = DeviceFlowState::Success(this_credential);
+            }
+        }
+    }
+}
+
+fn calculate_expiry(expires_in: i64) -> String {
+    let expires_in = Duration::seconds(expires_in);
+    let mut expiry: DateTime<Utc> = Utc::now();
+    expiry += expires_in;
+    expiry.to_rfc3339()
+}
+
+pub fn credential_error(msg: String) -> DeviceFlowError {
+    DeviceFlowError::GitHubError(msg)
+}
+
+pub async fn send_request(
+    device_flow: &mut DeviceFlow,
+    url: String,
+    body: String,
+) -> Option<HashMap<String, serde_json::Value>> {
+    let client = reqwest::Client::new();
+    let response_struct = client
+        .post(&url)
+        .header("Accept", "application/json")
+        .body(body)
+        .send();
+
+    match response_struct.await {
+        Ok(resp) => match resp.json::<HashMap<String, serde_json::Value>>().await {
+            Ok(hm) => Some(hm),
+            Err(err) => {
+                device_flow.state = DeviceFlowState::Failure(err.into());
+                None
+            },
+        },
+        Err(err) => {
+            device_flow.state = DeviceFlowState::Failure(err.into());
+            None
+        },
+    }
+}
+
+/// floxHub authentication commands
+#[derive(Clone, Debug, Bpaf)]
+pub enum Auth2 {
+    /// Login to floxhub (requires an existing github account)
+    #[bpaf(command)]
+    Login,
+}
+
+impl Auth2 {
+    pub async fn handle(self, _config: Config, _flox: Flox) -> Result<()> {
+        subcommand_metric!("auth2");
+        let mut client_id = String::new();
+        // TODO there is no obvious way to deal with
+        // identifying configuration that is not hard-coded into source
+        // feel free to suggest actionable alternatives that work in the existing
+        // cli codebase
+        client_id.push_str("Iv1.3b00a7bb5f910259");
```

Comment:
add to pkgs/flox/default.nix


```suggestion
        // TODO there is no obvious way to deal with
        // identifying configuration that is not hard-coded into source
        // feel free to suggest actionable alternatives that work in the existing
        // cli codebase
        let client_id = env!("OAUTH_CLIENT_ID").to_string();
```